src/clients/python/__init__.py

```diff
@@ -1501,6 +1504,7 @@ def async_run_with_cb(self, callback, inputs, outputs, batch_size=1, flags=0):
         wrapped_cb = partial(self._async_callback_wrapper, self._callback_resources_dict_id, callback)
         c_cb = _async_run_callback_prototype(wrapped_cb)
 
+        self._lock.acquire()
         # Run asynchronous inference...
```

Comment:
Use a try .. finally to make sure you always release the lock. Here is an exception is raised below the lock won't be released.

Reply:
using "with self._lock:" block to ensure the lock is released in all cases.